### PR TITLE
Override ParameterHandler.__getitem__

### LINKED ISSUE #599: Overriding `__getitem__` in ForceField
**Is your feature request related to a problem? Please describe.**
It would be nice to have cleaner lookups from `ForceField` objects to `ParameterHandler` objects, and possibly also directly to the individual parameters
```python3
>>> ff = ForceField('openff-1.1.0.offxml')
>>> ff['Bonds']
<openforcefield.typing.engines.smirnoff.parameters.BondHandler object at 0x7faeffb64b90>
>>> # Also could do
>>> # ff['Bonds']['b12']
# returns the particular parameter object
```

**Describe the solution you'd like**
This can be done by overriding `__getitem__` of `ForceField` and its individual `ParameterHandler` objects. Although some care is needed to ensure strange things don't happen. 

**Describe alternatives you've considered**
`ff.get_parameter_handler('Bonds')` works fine, as does `ff.get_parameter_handler('Bonds').get_parameter({'id': 'b12'})`, but the syntax sugar of `ff['Bonds']['b12']` would be preferred.

**Additional context**
`ParameterHandlers.get_parameter()` doesn't simply take in a value, it takes in a dict mapping that allows it to look up against SMIRKS patterns, ids, or in principle whatever other attribute is stored.

I saw this brought up in #310 and it has probably been mentioned elsewhere.

@j-wags says:

>Note, though, that ForceField.get_parameter_handler has a very important behavior that is different form a lookup -- If the parameterhandler for the requested SMIRNOFF section tag doesn't exist, it is initialized by get_parameter_handler, whereas I'd expect a lookup for the same  tag to fail.

### PR DESCRIPTION
**Pull request by mattwthompson**
Resolves  #599. Not a blocker for System work but it's really nice to be able to do `forcefield[handler_name][smirks]` to directly look up a parameter. The only lookup route is SMIRKS; other  types fail and strings that aren't parsed as SMIRKS (i.e. `id`) go nowhere.

- [x] Tag issue being addressed
- [x] Add [tests](https://github.com/openforcefield/openforcefield/tree/master/openforcefield/tests)
- [x] Update docstrings/[documentation](https://github.com/openforcefield/openforcefield/tree/master/docs), if applicable
- [x] [Lint](https://open-forcefield-toolkit.readthedocs.io/en/latest/developing.html#style-guide) codebase
- [x] Update [changelog](https://github.com/openforcefield/openforcefield/blob/master/docs/releasehistory.rst)

### CODE REVIEW

**Review by j-wags — CHANGES_REQUESTED** [review]
Is there any downside to consolidating all of this logic in `ParameterList.__getitem__`, and routing `ParameterHandler.__getitem__` straight to that? I think that would reduce complexity (and I'd be happy to make this change if you approve)

File: openforcefield/typing/engines/smirnoff/parameters.py [hunk 1]
```diff
@@ -2544,6 +2550,26 @@ def assign_partial_bond_orders_from_molecules(topology, pbo_mols):
...
+    def __getitem__(self, val):
+        """
+        Syntax sugar for lookikng up a ParameterType in a ParameterHandler
+        based on its SMIRKS.
+        """
+        if isinstance(val, str):
+            params = self.get_parameter({"smirks": val})
+            if len(params) == 1:
+                return params[0]
+            elif len(params) > 1:  # Can this ever be hit?
```
[the remark this answers is not in the document]
Comment (mattwthompson) [reply openforcefield/typing/engines/smirnoff/parameters.py]: Bumping this as the `ParameterList` implementation may or may not catch this in the same way. Far from blocking IMHO

**Comment by mattwthompson** [comment]
The `ParameterList.__getitem__` implementation seems better; simpler with less duplicated code, and exhibited similar behavior to what is in the new tests.

Note I changed some default exceptions to `ParameterLookupError` or something like that

**Review by j-wags — APPROVED** [review]
Awesome. Thanks for integrating my proposal -- I hope this reduces redundancy a bit!

Please list the `IndexError` --> `ParameterLookupError` change under `Behavior Changed` in the release notes, then this should be ready to merge!

**Pull request merged**